zipline/src/commonMain/kotlin/app/cash/zipline/internal/bridge/calls.kt

```diff
@@ -296,3 +297,78 @@ internal class ResultSerializer<T>(
     }
   }
 }
+
+/**
+ * Immediate result from invoking a suspending function, that may include either a value (if the
+ * call never suspended) or a cancel callback (if it did suspend).
+ */
+internal class SuspendingResult<T>(
+  /** The function suspended. Only non-null for suspend calls. */
+  val cancelCallback: CancelCallback? = null,
+
+  /** The function failed without suspending. */
+  val failure: Throwable? = null,
+
+  /** The function succeeded without suspending. */
+  val success: T? = null,
```

Comment:
Do you think we need an `init { }` which enforces the invariant that one (and only one) is set? Deserialization below could create an instance with all `null`s, for example.

Reply:
Good idea. Done.